polynote-server/src/main/scala/polynote/server/repository/fs/LocalFilesystem.scala

```diff
@@ -22,9 +23,55 @@ class LocalFilesystem(maxDepth: Int = 4) extends NotebookFilesystem {
     content <- readBytes(is).ensuring(ZIO.effectTotal(is.close()))
   } yield new String(content.toArray, StandardCharsets.UTF_8)
 
+  /**
+    * Writes the content to the given path "atomically" – at least, in a way where the existing file at that path (if
+    * present) won't be unrecoverable if writing fails for some reason.
+    */
+  private def writeAndSwap(path: Path, content: Array[Byte]) =
+    try {
+      // first, try creating the file in case it doesn't exist. If it doesn't, go ahead and write directly to it.
+      // it's not great to use an exception for normal flow control, but I don't think there's another way to avoid a
+      // race condition here (e.g. if we checked that the file doesn't exist, but then it does by the time we go to
+      // write to it)
+      Files.createFile(path)
+      Files.write(path, content)
+    } catch {
+      case _: FileAlreadyExistsException =>
+        // the file already exists, so we're going to write to a temporary file next to it and then swap it.
+        // this is to prevent unrecoverable incomplete writes if the application dies during the write.
+        val filename = path.getFileName.toString.stripPrefix(".")
+        val tmpPath = path.resolveSibling(s".$filename.swp")
```

Comment:
Should we attempt to recover old `swp` files? We're just overwriting them here, but I'm worried about the scenario when polynote is run as a service:
1. error occurs
2. polynote crashes, and is restarted by the system
3. user sees the crash/disconnect, reloads the page

what does the user see? will they lose the swap file? what happens if they reload multiple times (this will be triggered every time the file is opened/closed right?)

or, are the swp files purely a mechanism to avoid truncation in which case we don't cate about them at all?

Reply:
I *think* that by the time you go to create a new swap file, it means you have a good notebook to write. So it should be fine to discard old ones.

Reply:
I guess I'm worried the swp file might have the delta of what you were writing but hadn't yet persisted. When it goes back to create the new swap file after restarting, the server will be have a good notebook based on `notebook.ipynb`, not `notebook.ipynb.swp`.